### ISSUE TITLE
Support For Hamming Distance

### ISSUE THREAD

**Issue by seancarroll** [pgvector/pgvector#429]
Would be interested to see pgvector support hamming distance. An example of an existing implementation can be found in [lantern](https://github.com/lanterndata/lantern)

Example Use Case: Storing PDQ hashes, a photo-hashing algorithm, as binary vectors which can be compared via hamming distance.

**Comment by ankane** [comment]
Hi @seancarroll, thanks for the suggestion. Added an initial version (for exact search) in the [hamming-distance branch](https://github.com/pgvector/pgvector/compare/hamming-distance).

**Comment by ankane** [comment]
It looks like it's currently possible to do exact search with bit strings (without pgvector), fwiw.

```tsql
CREATE FUNCTION hamming_distance(a varbit, b varbit) RETURNS float8
    LANGUAGE SQL IMMUTABLE STRICT PARALLEL SAFE
    RETURN bit_count(a # b);

SELECT hamming_distance(B'1111111100000000', B'1111000000001111');
```

Indexing will take a bit more work (to index non-vector types, but that's already planned), but would like to see how much demand there is for this first.

**Comment by seancarroll** [comment]
Hi @ankane thanks for the follow up. Yea I'm interested in the indexing use case. For the use case I was considering it would be indexing a binary vector but good to know indexing on non-vector types is planned.

**Comment by ankane** [comment]
It might be worth trying exact search depending on the size of your data.

```sql
CREATE TABLE items (id bigserial PRIMARY KEY, image_hash bit(256));
INSERT INTO items (image_hash) SELECT i::bit(256) FROM generate_series(1, 1000000) i;
SELECT id FROM items ORDER BY bit_count(image_hash # 500000::bit(256)) LIMIT 10;
```

The `SELECT` query completes in under 70 ms on my machine with 1 million hashes.

**Comment by seancarroll** [comment]
Sorry I should have been clearer in the description of the use case. What you have certainly works great for an exact search but the target use case would be for a similarity search. One of the interesting features of PDQ hashes is that perceptually similar images produce similar outputs and their similarity can be determined via hamming distance.

So the use case would be indexing a set of images and then being able to do a similarity query to do something like give me N number of results within some threshold distance D from this vector

**Comment by ankane** [comment]
The above query does similarity search for Hamming distance on 256 bit vectors (which is the hash size produced by PDQ). It gets the 10 nearest neighbors, but you could also add a `WHERE` clause to filter by distance as well.

```tsql
SELECT id FROM items
    WHERE bit_count(image_hash # 500000::bit(256)) <= 10  -- threshold distance D
    ORDER BY bit_count(image_hash # 500000::bit(256))     -- similarity search
    LIMIT 100;                                            -- N number of results
```

**Comment by seancarroll** [comment]
You're absolutely right. I think an index might still be valuable but I'm no expert. Providing crude numbers on my mac 1m returns within 300 ms or so while 10mil takes roughly 4.5sec. Granted there are likely ways to get some perf boot. Anywho, I appreciate all the details you provided

**Comment by dannyseismic** [comment]
@ankane I'm not sure how you measure demand, but this is also something we're interested in for the same use case (image hash similarity search).

**Comment by pshrest2** [comment]
@ankane this is something I am very interested in as well.

**Comment by a-mckinley** [comment]
Binary text embedding models appear to have had a major uptick in capability recently. Index support for hamming distance search would be a great major step

**Comment by ankane** [comment]
Support for HNSW indexing is available in the `bitvector` branch for anyone who wants to try it (in a non-production environment).

```sql
CREATE TABLE items (id bigserial PRIMARY KEY, embedding bit(3));
INSERT INTO items (embedding) VALUES (B'000'), (B'111');
CREATE INDEX ON items USING hnsw (embedding bit_hamming_ops);
SELECT * FROM items ORDER BY embedding <~> B'101' LIMIT 5;
```

**Comment by ankane** [comment]
This has been merged (94a444f02958f04946fc2dfccf0e10fe3718b660) and will be included in 0.7.0 (#508). Thanks @seancarroll for the suggestion!
